File: Extract_b1b2b3_position.m
function Extract_b1b2b3_position(a,b)
global SA param player

disp('Start identifying B1B2B3 Positions ...')

hsl = findobj('Tag', 'uitable_shotlist');
sz = size(hsl.Data);

for ti = 1:sz(1)
    identify_ShotID(ti, hsl.Data, hsl.ColumnName)
    si = SA.Current_si;

    if SA.Table.Interpreted{si} == 0
        
        [b1b2b3, b1i, b2i, b3i] = str2num_B1B2B3(SA.Table.B1B2B3{si});
        
        % Create fields
        for bi = 1:3
            
            % Make calculation
            % ===============EDIT THIS HERE ===============
            % First check is column already available, then update visibility
            
            %% PosX
            varname = ['B',num2str(bi),'posX'];
            ind = find(strcmp(SA.Table.Properties.VariableNames, varname));
            SA.Table.(varname){si} = SA.Shot(si).Route(b1b2b3(bi)).x(1)/param.size(2)*8;
            
            
            %% PosY
            varname = ['B',num2str(bi),'posY'];
            % First check is column already available, then update visibility
            ind = find(strcmp(SA.Table.Properties.VariableNames, varname));
            SA.Table.(varname){si} = SA.Shot(si).Route(b1b2b3(bi)).y(1)/param.size(1)*4;
            
            err.code = [];
            err.text = [];
            
            % Update the err
            % First check is column already available, then update visibility
            ind = find(strcmp(SA.Table.Properties.VariableNames, 'ErrorID'));
            SA.Table.ErrorID{si} = err.code;
            
            ind = find(strcmp(SA.Table.Properties.VariableNames, 'ErrorText'));
            SA.Table.ErrorText{si} = err.text;
            
        end
    end
end
% update GUI
update_ShotList
player.uptodate = 0;
PlayerFunction('plotcurrent',[])

disp(['done (',mfilename,')'])

File: extract_b1b2b3_position.py
from player_function import player_function
from str2num_b1b2b3 import str2num_b1b2b3

def extract_b1b2b3_position(SA, param):
    print(f"Start ({__name__})")
    for bi in range(0, 3):
        SA[f"B{bi+1}posX"] = None  # Initialize columns for ball positions
        SA[f"B{bi+1}posY"] = None

    for si in range(len(SA)):
        b1b2b3, b1i, b2i, b3i = str2num_b1b2b3(SA.iloc[si]['B1B2B3'])

        for bi in range(0, 3):
            varname_x = f"B{bi+1}posX"
            SA.at[si, varname_x] = SA.iloc[si][f'Route{b1b2b3[bi]}']['x'][0] / param['size'][1] * 8

            varname_y = f"B{bi+1}posY"
            SA.at[si, varname_y] = SA.iloc[si][f'Route{b1b2b3[bi]}']['y'][0] / param['size'][0] * 4

        SA.at[si, 'ErrorID'] = None
        SA.at[si, 'ErrorText'] = None

    print(f"done ({__name__})")
    return SA
